src/Compiler/Driver/GraphChecking/Types.fs

```diff
@@ -146,20 +151,25 @@ type internal QueryTrie = LongIdentifier -> QueryTrieNodeResult
 
 /// Helper class to help map signature files to implementation files and vice versa.
 type internal FilePairMap(files: FileInProject array) =
+    let sigFiles = files |> Array.filter _.ParsedInput.IsSigFile
+    let implFiles = files |> Array.filter _.ParsedInput.IsImplFile
+
     let buildBiDirectionalMaps pairs =
         Map.ofArray pairs, Map.ofArray (pairs |> Array.map (fun (a, b) -> (b, a)))
 
-    let implToSig, sigToImpl =
-        files
-        |> Array.choose (fun f ->
-            match f.ParsedInput with
-            | ParsedInput.SigFile _ ->
-                files
-                |> Array.skip (f.Idx + 1)
-                |> Array.tryFind (fun (implFile: FileInProject) -> $"{implFile.FileName}i" = f.FileName)
-                |> Option.map (fun (implFile: FileInProject) -> (implFile.Idx, f.Idx))
-            | ParsedInput.ImplFile _ -> None)
-        |> buildBiDirectionalMaps
+    let pairs =
+        sigFiles
+        |> Array.map (fun sigFile ->
+            implFiles
+            |> Array.tryFind (fun (implFile: FileInProject) -> $"{implFile.FileName}i" = sigFile.FileName)
```

Comment:
Luckily all of this is a no-op of sigFiles is empty (= most projects).

It does O(#sigFiles * #implFiles) new string allocations (for the interpolated string).

Do we need to be that liberate for the order?
I would have assumed that a sigFile can just look at idx+1, idx-1  in the overall files array;; and then  compare spans of characters. Or are there test cases with more relaxed rules on the ordering?

We should aim to be 
- no-op if no sig files
- linear if sig files present but correctly ordered
- and only worse for degenerate orderings.


(for regular compilation it does not matter, but design time scenario via transparent compiler is IIRC making use of the graph)

Reply:
Yes, I'm not sure why it was like this instead of just comparing `QualifiedNameOfFile.Id`.

But this covers one edge case I was not aware of: fsi file does not need to directly precede the implementation, it just needs to come before. I can't remember out of my head but there are some public projects in the wild that do it that way.